Source/WebCore/dom/ContainerNode.cpp

```diff
@@ -86,11 +86,29 @@ unsigned ScriptDisallowedScope::s_count = 0;
 ScriptDisallowedScope::EventAllowedScope* ScriptDisallowedScope::EventAllowedScope::s_currentScope = nullptr;
 #endif
 
-ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertion(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> DidRemoveElements
+ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertionMaybeAsync(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> RemoveAllChildrenResult
+{
+    // FIXME: remove this when other platforms support Opportunistic Task Scheduler
+#if PLATFORM(MAC)
+    // Keep node deletion synchronous if removing all children will delay destroying the document
+    if (isDocumentNode() || isDocumentFragment())
+        return removeAllChildrenWithScriptAssertion(source, children, deferChildrenChanged);
+    auto removeAllChildrenResult = removeAllChildrenWithScriptAssertion(source, children, deferChildrenChanged);
+    if (removeAllChildrenResult.canBeDelayed == CanDelayNodeDeletion::Yes)
+        document().asyncNodeDeletionQueue().addIfSubtreeSizeIsUnderLimit(WTFMove(children), removeAllChildrenResult.subTreeSize);
+    return removeAllChildrenResult;
+#else
+        return removeAllChildrenWithScriptAssertion(source, children, deferChildrenChanged);
+#endif
+}
+
+ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertion(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> RemoveAllChildrenResult
 {
     ASSERT(children.isEmpty());
     collectChildNodes(*this, children);
 
+    auto canDelayNodeDeletion = CanDelayNodeDeletion::Yes;
```

Comment:
It seems that we can declare this variable much later at where treeSize is declared.